blockchain/state_manager/src/lib.rs

```diff
@@ -294,50 +298,76 @@ where
         epoch: ChainEpoch,
         rand: &R,
         base_fee: BigInt,
-        callback: Option<CB>,
+        mut callback: Option<CB>,
         tipset: &Arc<Tipset>,
     ) -> Result<CidPair, Box<dyn StdError>>
     where
-        R: Rand,
+        R: Rand + Clone + 'static,
         V: ProofVerifier,
         CB: FnMut(&Cid, &ChainMessage, &ApplyRet) -> Result<(), String>,
     {
         let db = self.blockstore_cloned();
-        let mut buf_store = Arc::new(BufferedBlockStore::new(db.as_ref()));
-        let store = buf_store.as_ref();
         let lb_wrapper = SMLookbackWrapper {
             sm: self,
-            store,
+            store: self.blockstore(),
             tipset,
             verifier: PhantomData::<V>::default(),
         };
 
-        let mut vm = VM::<_, _, _, _, _, V>::new(
-            p_state,
-            store,
-            epoch,
-            rand,
-            base_fee,
-            get_network_version_default,
-            &self.genesis_info,
-            &lb_wrapper,
-        )?;
+        let rand_clone = rand.clone();
+        let create_vm = |state_root, epoch| {
+            VM::<_, _, _, _, V>::new(
+                state_root,
+                db.as_ref(),
+                db.clone(),
+                epoch,
+                &rand_clone,
+                base_fee.clone(),
+                get_network_version_default(epoch),
+                self.genesis_info.clone(),
+                &lb_wrapper,
+                self.engine.clone(),
+            )
+        };
+
+        let mut parent_state = *p_state;
+
+        for epoch_i in parent_epoch..epoch {
+            if epoch_i > parent_epoch {
+                let mut vm = create_vm(parent_state, epoch_i)?;
+                // run cron for null rounds if any
+                if let Err(e) = vm.run_cron(epoch_i, callback.as_mut()) {
+                    log::error!("Beginning of epoch cron failed to run: {}", e);
+                }
+
+                parent_state = vm.flush()?;
+            }
+
+            if epoch_i == networks::UPGRADE_ACTORS_V4_HEIGHT {
```

Comment:
should this check against any of the actor upgrade heights?

Reply:
Right now, forest doesn't support upgrades other than the one at this epoch. This PR drops all support for network upgrades and replaces them with panics. Issue #1454 and #1452 have been created to figure out a consistent way of dealing with network upgrades all the way back to genesis.